File: src/golang/basic/deque.go
package basic

import "log"

type Deque struct{
	items	[]interface{}
	size	uint64
}

func NewDeque() *Deque {
	deque := &Deque{
		items: make([]interface{}, 0),
		size: 0,
	}

	return deque
}

func (d *Deque) AddFirst(item interface{}) {
	if item == nil {
		log.Fatal("argument for AddFirst() should not nil")
	}
	if d.size == 0 {
		d.items = append(d.items, item)
		d.size++
	}else{
		tmp := make([]interface{}, 0)
		tmp = append(tmp, item)
		d.items = append(tmp, d.items...)
		d.size++
	}
}

func (d *Deque) AddLast(item interface{}) {
	if item == nil {
		log.Fatal("argument to AddLast() should not nil")
	}
	d.items = append(d.items, item)
	d.size++
}

func (d *Deque) RemoveFirst() interface{} {
	if d.size == 0 {
		log.Fatal("deque is empty")
	}
	item := d.items[0]
	d.items = d.items[1:]
	d.size--
	return item
}

func (d *Deque) RemoveLast() interface{} {
	if d.size == 0 {
		log.Fatal("deque is empty")
	}
	item := d.items[d.size-1]
	d.items = d.items[:(d.size-1)]
	d.size--
	return item

}

func (d *Deque) Size() uint64 {
	return d.size
}

func (d *Deque) IsEmpty() bool {
	return d.size == 0
}

File: src/python/basic/deque.py
# -*- coding: utf-8 -*-

class Deque(object):
    def __init__(self):
        self.items = []
        self._size = 0

    def add_first(self, item):
        self.items.insert(0, item)
        self._size += 1

    def add_last(self, item):
        self.items.append(item)
        self._size += 1

    def remove_first(self):
        if self.is_empty():
            raise Exception('deque is empty')
        item = self.items.pop(0)
        self._size -= 1
        return item

    def remove_last(self):
        if self.is_empty():
            raise Exception('deque is empty')
        item = self.items.pop()
        self._size -= 1
        return item

    def is_empty(self):
        return self._size == 0

    def size(self):
        return self._size

def test_deque():
    d = Deque()
    d.add_first(1)
    d.add_first(2)
    d.add_last("a")
    d.add_first("b") # ["b", 2,1,"a"]

    assert d.size() == 4
    assert d.remove_first() == "b"
    assert d.remove_last() == "a"
